### ISSUE TITLE
reload .envrc before executing command

### ISSUE THREAD

**Issue by nauni77** [direnv/direnv#907]
If a second bash/ shell (B) will change the `.envrc` this change is not used at first command in bash/ chell (A).

Open Bash A (at home of user):
```
$ echo "export FOO=myFoo" > .envrc
direnv: loading ~/.envrc
direnv: export +FOO
$ echo $FOO
myFoo
```

Open Bash B  (at home of user):
```
$ echo $FOO
myFoo
$ echo "export FOO=myFoo2" > .envrc
direnv: loading ~/.envrc
direnv: export +FOO
$ echo $FOO
myFoo2
```

Now use Bash A again and you will see the first command use the wrong/ old content:
```
$ echo $FOO
myFoo
direnv: loading ~/.envrc
direnv: export +FOO
$ echo $FOO
myFoo2
$
```

This means the first command will be executed with the wrong settings. How can I achieve that even the first command after change will return `myFoo2` instead of `myFoo`.

Thanks in advance and best regards, 
Oli

**Comment by zimbatm** [comment]
It's a valid observation but I don't really know if there is a way to fix it.

direnv basically depends on callbacks that are provided by the shell to update the environment. It installs itself to be called before a new prompt is being displayed.

In this case, we would need bash to provide a hook to execute direnv *before* running the command. I don't think that this facility exists.

**Comment by nauni77** [comment]
> In this case, we would need bash to provide a hook to execute direnv _before_ running the command. I don't think that this facility exists.

That's true - even I don't know if this is possible. But it would be really great if it is possible. Maybe someone who reads this will have an idear to solve this? Thanks in advance!

For example I use this to configure the cluster access and the `.envrc` is located at my users home. And I always have to take care if I use more than one terminal tab.

**Comment by untitaker** [comment]
I just stumbled across this, you can probably work around this in your envrc by patching custom code into the user's PS1, from within your envrc:

```
function reload_variables() {
    ...
}
PS1+='`reload_variables`'
export PS1
```

it is just really hideous and you need to absolutely make sure that function doesn't output anything

**Comment by zeptonaut** [comment]
The most surprising thing here is that if you have multiple terminals open and a reload happens in one of them (either executed manually via `direnv reload .` or executed alongside a `git checkout`, the same reload doesn't happen in the other terminals open in the same directory.

I see a few options:
- After doing some research here, it does look like `bash-preexec` has managed to implement a `preexec` hook for `bash` in much the same way that one already exists for `zsh` and `tcsh`. `direnv` could try to do something similar, or just leverage `bash-preexec` somehow. It seems like direnv already has some notion of a "dirty" state of the environment because it knows to reload the environment when the next shell is printed, so it seems like this preexec should only add minimal overhead unless the environment is indeed dirty and needs to be reloaded
- I wonder if signals could be used here to any effect. Each terminal running `direnv` could receive a signal (e.g. `SIGUSR1`) that tells it to reload the environment whenever one shell reloads its environment. That way, the shells could stay in sync.
- Worst comes to worst, `direnv` could just provide a `preexec` hook for shells where it's already supported to help with this.

A few false paths that looked promising but weren't:
- There's a `periodic` function that's also available in `zsh` and `tcsh` that allows shells to periodically run commands every X seconds while they're idle. I don't think this buys us anything because the only shells where this `periodic` command is available are shells that already have a `preexec` command, which is probably a better fit for what we want anyhow.
- `SIGHUP` looked like a promising option because it's often co-opted by daemon processes as a signal for reloading configuration changes (similar to what we want), but because it is actually possible that `direnv` would receive a `SIGHUP`, I don't think it's a good fit for us. (Daemon processes specifically coopt this signal because there's no real case where they'd receive one under normal circumstances.)
- I tried doing the naive thing and implementing my own reload hook via a `zsh` `preexec` hook, e.g.

```zsh
preexec() {
    direnv reload .;
    echo "DIRENV RELOADING!";
}
```

When I did this, I did indeed see `DIRENV_RELOADING!` printed, but the command was executed with the same stale environment that it would have been before. Maybe this suggests that co-opting the preexec hook for our purposes isn't the catch-all that I hoped it would be. (Maybe that preexec hook is itself run in a subshell, so it's not possible to affect the direnv subshell from it? I'm not sure.)
